### PR TITLE
Add `PuppetJS` and `PuppetDOM` constructors support with tests

### PR DESCRIPTION
**Pull request by alshakero**
In this PR I:

1) Added `puppet.js` and `puppet-dom.js` files to `src` directory. But added a deprecation warning in both files.
2) Now, in `palindrom.js` both `new Palindrom()` and `new PuppetJS()` work. I added a spec for this.
2) Now, in `palindrom-dom.js` both `new PalindromDOM()` and `new PuppetDOM()` work. I added a spec for this.
3) In case of using old `puppet-redirect`. Now `PalindromDOM` supports both events `palindrom-redirect-pushstate` and `puppet-redirect-pushstate`. But it emits a warning every time it handles `puppet-redirect-pushstate`. I added specs to test both events.
4) I tested both `puppet.js` and `puppet-dom.js` for warnings in KitchenSink and they both work and emit warnings.
5) I tested both `palindrome.js` and `palindrome-dom.js` in KitchenSink for backward compatibility and they both work and emit warnings.

P.S: failing Travis isn't relevant to the edits.

### CODE REVIEW

**Comment by tomalec** [comment]
I like the idea of being as backward compatible as possible. But I think (1.) keeping copies of source in `puppet.js` is too much to maintain. I believe in most of our use cases somebody will either use old PuppetJs 2.4.0 and `puppet.js`, or new Palindrom 3.0.0 and `palindrom.js`. I think aliasing constructor with `Puppet`/`PuppetDOM` + warning should be enough.

404 for `puppet.js` fails fast, so I believe any developer would notice and fix it quickly.

**Comment by alshakero** [comment]
@tomalec I totally agree, I was just trying to make the most of it. Keeping the new files only is cleaner. 
Should bump to 3.0.0 and release when I finish this?

**Comment by tomalec** [comment]
> Should bump to 3.0.0 and release when I finish this?

Let's wait for 
https://github.com/Palindrom/Palindrom/issues/117#issuecomment-283370844

File: src/palindrom-dom.js [hunk 1]
```diff
@@ -10,29 +10,43 @@
    * @param {Object} [options] map of arguments. See README.md for description
    */
   var PalindromDOM = function (options){
-    options || (options={});
+    if (typeof options !== "object") {
+            throw new Error("'options' is not an object");
```
Comment (tomalec) [inline src/palindrom-dom.js]: I'd either do more descriptive message like "Palindorm() requires and object argument...." as by running `new Palindrom()` you have no idea what are the `options`. Or  skip it totally, and reduce it to 
`if(!(options && options.remoteUrl)) { throw new Error('remoteUrl option is required');`

File: src/palindrom-dom.js [hunk 2]
```diff
@@ -10,29 +10,43 @@
...
+      }
+    if (!options.remoteUrl) {
+          throw new Error('remoteUrl is required');
+    }
     var onDataReady = options.callback;
     this.element = options.listenTo || document.body;
     var clickHandler = this.clickHandler.bind(this);
     this.historyHandler = this.historyHandler.bind(this);
+    this.historyHandlerDeprecatedMessage = function () {
```
Comment (tomalec) [inline src/palindrom-dom.js]: by the name I thought it's the mesage for the handler thsat would be called anyways, WDYT about `historyHandlerDeprecated` ?

File: src/palindrom-dom.js [hunk 3]
```diff
@@ -10,29 +10,43 @@
...
       this.listen();
       onDataReady && onDataReady.call(this, obj);
     };
     this.listen = function(){
       this.listening = true;
-
       this.element.addEventListener('click', clickHandler);
       window.addEventListener('popstate', this.historyHandler); //better here than in constructor, because Chrome triggers popstate on page load
-      this.element.addEventListener('palindrom-redirect-pushstate', this.historyHandler);
+      this.element.addEventListener('palindrom-redirect-pushstate', this.historyHandler.bind(this));
```
Comment (tomalec) [inline src/palindrom-dom.js]: I think `.bind(this)`  is not needed here.

File: src/palindrom-dom.js [hunk 4]
```diff
@@ -10,29 +10,43 @@
...
+      
+      /* backward compatibility: for people using old puppet-redirect */
+      this.element.addEventListener('puppet-redirect-pushstate', this.historyHandlerDeprecatedMessage.bind(this));
```
Comment (tomalec) [inline src/palindrom-dom.js]: `.bind(this)` coulb be not needed

File: src/palindrom-dom.js [hunk 5]
```diff
@@ -10,29 +10,43 @@
...
+    if (!options.remoteUrl) {
+          throw new Error('remoteUrl is required');
+    }
     var onDataReady = options.callback;
     this.element = options.listenTo || document.body;
     var clickHandler = this.clickHandler.bind(this);
     this.historyHandler = this.historyHandler.bind(this);
+    this.historyHandlerDeprecatedMessage = function () {
+      console.warn("`puppet-redirect-pushstate` event is deprecated, please use `palindrom-redirect-pushstate`, if you're using `puppet-redirect`, please upgrade to `palindrom-redirect`");
+      this.historyHandler.bind(this)();
```
Comment (tomalec) [inline src/palindrom-dom.js]: I'd rather move this `.bind(this)` below, not to create new function every time it's called, or just use closure.

File: src/palindrom.js [hunk 6]
```diff
@@ -514,12 +514,18 @@
...
 
+
   /**
    * Defines a connection to a remote PATCH server, serves an object that is persistent between browser and server.
    * @param {Object} [options] map of arguments. See README.md for description
    */
-  function Palindrom(options) {
-    options || (options={});
+  function Palindrom(options) {    
+    if (typeof options !== "object") {
```
Comment (tomalec) [inline src/palindrom.js]: see `palindor-dom.js` comments

**Pull request merged**